### PR TITLE
topology_custom/test_tablets: add remove/replace tests for edge cases

### PR DESCRIPTION
**Pull request by bhalevy**
Test cases related to #21826:

1. test_remove_failure_with_no_normal_token_owners_in_dc: attempts to
    remove a node with another node down in the datacenter, leaving
    no normal token owners in that dc (reproducing #21826).
    Removenode is expected to fail in this case since it
    should have no place to rebuild the removed node replicas,
    yet it currently succeeds unexpectedly.

2. test_remove_failure_then_replace: verify that removenode
    fails as expected when there are not enough nodes to
    rebuild its replicas on, with and without additional zero-token nodes.

3. test_replace_with_no_normal_token_owners_in_dc: verify that
    nodes can be replaced in a datacenter that has no live
    token owners, with and without additional zero-token nodes.
    Tablet replace uses all replicas to rebuild the lost replicas
    and therefore should succeed in the edge case.
    The restored data is verified as well.

Refs #21826

* New tests, no backport needed

### CODE REVIEW

**Comment by bhalevy** [comment]
In v2 ([6e96cdd](https://github.com/scylladb/scylladb/commit/6e96cdd8544b6b5ea432b153ddb14105cf6def9a)):
- test: pylib: _cluster_remove_node: mark server as removed only when removenode succeeded
- test: pylib: _cluster_remove_node: log message on successful paths
- test: topology_custom: add test_remove_failure_then_replace_with_zero_token_node
  - verify replace after failed removenode

**Comment by bhalevy** [comment]
rebased

**Comment by bhalevy** [comment]
@kbr-scylla can you please review too?

File: test/topology_custom/test_tablets.py [hunk 1]
```diff
@@ -626,3 +627,35 @@ async def test_orphaned_sstables_on_startup(manager: ManagerClient):
...
+    num_keys = 1000
+    await asyncio.gather(*[cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({k}, {k%3});") for k in range(num_keys)])
+
+    node_to_remove = servers['dc1'][0]
+    initiator_node = servers['dc1'][1]
+
+    logger.info("Attempting removenode - expected to fail")
+    await manager.server_stop_gracefully(node_to_remove.server_id)
+    await manager.remove_node(initiator_node.server_id, server_id=node_to_remove.server_id,
+                              expected_error="Removenode failed. See earlier errors (Rolled back: Failed to drain tablets: std::runtime_error (Unable to find new replica for tablet")
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply test/topology_custom/test_tablets.py:656]: Done in next version

**Comment by bhalevy** [comment]
In v3 ([e591d67](https://github.com/scylladb/scylladb/commit/e591d67e36c86e3ac1b7e15b316d5cee7cb001fa)):
- changed initial topology to have an additional zero-token node in dc1
- changed rf to `{'dc1': 2, 'dc2': 1}` although tablets should never migrate between DCs, let's strengthen the tests in this area.

File: test/topology_custom/test_tablets.py [hunk 2]
```diff
@@ -626,3 +627,34 @@ async def test_orphaned_sstables_on_startup(manager: ManagerClient):
...
+    servers['dc1'] = await manager.servers_add(servers_num=2, property_file={'dc': 'dc1', 'rack': 'rack1'})
+    servers['dc1'].append(await manager.servers_add(servers_num=1, config={'join_ring': False}, property_file={'dc': 'dc1', 'rack': 'rack1'}))
+    servers['dc2'] = await manager.servers_add(servers_num=2, property_file={'dc': 'dc2', 'rack': 'rack2'})
+    servers['dc3'] = await manager.servers_add(servers_num=1, config={'join_ring': False}, property_file={'dc': 'dc3', 'rack': 'rack3'})
+
+    cql = manager.get_cql()
+    await cql.run_async(f"CREATE KEYSPACE test WITH replication = {{ 'class': 'NetworkTopologyStrategy', 'dc1': 2, 'dc2': 1 }} AND tablets = {{ 'initial': 1 }}")
+    await cql.run_async("CREATE TABLE test.test (pk int PRIMARY KEY, c int);")
+    num_keys = 1000
+    await asyncio.gather(*[cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({k}, {k%3});") for k in range(num_keys)])
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply test/topology_custom/test_tablets.py]: dropped in v3

**Comment by bhalevy** [comment]
In v3 (b953120):
- Added more test cases
- In particular, reproducer for #21826 (currently marked as xfail)

**Review by tgrabiec — APPROVED** [review]

**Pull request closed without merging**